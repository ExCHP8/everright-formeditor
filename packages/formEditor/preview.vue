<script>
import { defineProps, ref, reactive, computed, provide, getCurrentInstance, watch, nextTick, onMounted } from 'vue'
import CanvesPanel from '@ER/formEditor/components/Panels/Canves'
import hooks from '@ER/hooks'
import utils from '@ER/utils'
import _ from 'lodash-es'
import defaultProps from './defaultProps'
import { ElMessage } from 'element-plus'
import { showNotify } from 'vant'
export default {
  name: 'Everright-form-preview'
}
</script>
<script setup>
const emit = defineEmits(['listener'])
const props = defineProps(defaultProps)
const layout = {
  pc: [],
  mobile: []
}
const state = reactive({
  store: [],
  selection: {},
  platform: utils.isPc() ? 'pc' : 'mobile',
  mode: 'preview',
  config: {},
  Namespace: 'formEditor',
  validateStates: [],
  data: {},
  fields: [],
  logic: {},
  fieldsLogicState: new Map()
})
const ns = hooks.useNamespace('Main', state.Namespace)
hooks.useLogic(state)
const getData = () => {
  const result = {}
  state.fields.forEach(e => {
    if (e.type === 'subform') {
      result[e.key] = e.list.map(e => {
        const cur = {}
        const children = []
        e.forEach(e => {
          e.columns.forEach(e => {
            children.push(e)
          })
        })
        children.forEach(e => {
          cur[e.key] = e.options.defaultValue
        })
        return cur
      })
    } else {
      if (!utils.checkIsInSubform(e)) {
        result[e.key] = e.options.defaultValue
      }
    }
  })
  return _.cloneDeep(result)
}
const fireEvent = (type, data) => {
  emit('listener', {
    type,
    data
  })
}
provide('Everright', {
  state,
  getData,
  props,
  fireEvent
  // checkFieldsValidation
})
const setData2 = (data, value) => {
  const newData = _.cloneDeep(data)
  layout.pc = newData.layout.pc
  layout.mobile = newData.layout.mobile
  state.store = newData.list
  state.fields = newData.fields
  const curLayout = _.cloneDeep(newData.layout[state.platform])
  utils.combinationData2(curLayout, state.fields)
  state.store = curLayout
  state.config = newData.config
  state.data = newData.data
  state.logic = newData.logic
  state.store.forEach((e) => {
    utils.addContext(e, state.store, false)
  })
  if (!_.isEmpty(value)) {
    state.fields.forEach((e) => {
      if (e.type === 'time' && !e.options.valueFormat) {
        e.options.valueFormat = 'HH:mm:ss'
      }
      if (value[e.key]) {
        e.options.defaultValue = value[e.key]
      }
    })
  }
}
const setValue = (field, value) => {
  if (field.type === 'time' && !field.options.valueFormat) {
    field.options.valueFormat = 'HH:mm:ss'
  }
  field.options.defaultValue = value
}
const setData1 = async (data, value) => {
  if (_.isEmpty(data)) return false
  const newData = utils.combinationData1(_.cloneDeep(data))
  state.store = newData.list
  state.config = newData.config
  state.data = newData.data
  state.fields = newData.fields
  state.logic = newData.logic
  state.store.forEach((e) => {
    utils.addContext(e, state.store)
  })
  // For SubformLayout.jsx to get the first data
  await nextTick()
  if (!_.isEmpty(value)) {
    state.fields.forEach((field) => {
      if (field.type === 'subform') {
        const addData = _.cloneDeep(field.list[0])
        for (let i = 1; i < value[field.key].length; i++) {
          field.list.push(addData)
        }
        field.list[field.list.length - 1].forEach(e => {
          utils.addContext(e, field)
        })
        field.list.forEach((e, index) => {
          e.forEach(e => {
            e.columns.forEach(e => {
              if (value[field.key]) {
                setValue(e, value[field.key][index][e.key])
              }
            })
          })
        })
      } else {
        if (!utils.checkIsInSubform(field)) {
          if (value[field.key]) {
            setValue(field, value[field.key])
          }
        }
      }
    })
  }
}
const setData = props.layoutType === 1 ? setData1 : setData2
defineExpose({
  switchPlatform (platform) {
    state.platform = platform
  },
  setData,
  getData
})
</script>
<template>
  <CanvesPanel v-if="state.store.length"></CanvesPanel>
</template>
