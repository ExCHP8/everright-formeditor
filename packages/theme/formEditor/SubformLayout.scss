@include b(SubformLayout) {
  @include e(content) {
    width: 100%;
  }
  .el-form-item {
    display: block;
    .el-form-item__label {
      display: block;
      height: auto;
      text-align: left;
      margin-bottom: 8px;
      line-height: 22px;
    }
  }
}
