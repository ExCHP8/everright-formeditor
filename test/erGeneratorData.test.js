import { describe, assert, expect, test, beforeEach } from 'vitest'
import { erFormConfig, erGeneratorData, erComponentsConfig, utils } from '@ER/formEditor'
describe('Generate field data:', () => {
  test('Email', () => {
    expect(erGeneratorData(erComponentsConfig.fieldsConfig[0].list[0], true, 'en')).toMatchSnapshot()
  })
  test('Cellphone', () => {
    expect(erGeneratorData(erComponentsConfig.fieldsConfig[0].list[1], true, 'en')).toMatchSnapshot()
  })
  test('URL', () => {
    expect(erGeneratorData(erComponentsConfig.fieldsConfig[0].list[2], true, 'en')).toMatchSnapshot()
  })
  test('Region', () => {
    expect(erGeneratorData(erComponentsConfig.fieldsConfig[0].list[3], true, 'en')).toMatchSnapshot()
  })
  test('ID number', () => {
    expect(erGeneratorData(erComponentsConfig.fieldsConfig[0].list[4], true, 'en')).toMatchSnapshot()
  })
  test('Input', () => {
    expect(erGeneratorData(erComponentsConfig.fieldsConfig[1].list[0], true, 'en')).toMatchSnapshot()
  })
  test('Textarea', () => {
    expect(erGeneratorData(erComponentsConfig.fieldsConfig[1].list[1], true, 'en')).toMatchSnapshot()
  })
  test('Number', () => {
    expect(erGeneratorData(erComponentsConfig.fieldsConfig[1].list[2], true, 'en')).toMatchSnapshot()
  })
  test('Radio', () => {
    const result = erGeneratorData(erComponentsConfig.fieldsConfig[1].list[3], true, 'en')
    result.columns[0].options.data = utils.generateOptions(3).map((e, i) => {
      e.label += i + 1
      return e
    })
    expect(result).toMatchSnapshot()
  })
  test('Checkbox', () => {
    const result = erGeneratorData(erComponentsConfig.fieldsConfig[1].list[4], true, 'en')
    result.columns[0].options.data = utils.generateOptions(3).map((e, i) => {
      e.label += i + 1
      return e
    })
    expect(result).toMatchSnapshot()
  })
  test('Select', () => {
    const result = erGeneratorData(erComponentsConfig.fieldsConfig[1].list[5], true, 'en')
    result.columns[0].options.data = utils.generateOptions(3).map((e, i) => {
      e.label += i + 1
      return e
    })
    expect(result).toMatchSnapshot()
  })
  test('Time', () => {
    expect(erGeneratorData(erComponentsConfig.fieldsConfig[1].list[6], true, 'en')).toMatchSnapshot()
  })
})
