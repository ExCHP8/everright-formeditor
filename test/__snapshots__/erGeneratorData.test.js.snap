// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generate field data: > Cascader 1`] = `
{
  "columns": [
    {
      "icon": "cascader",
      "id": "{{test-id-nanoid}}",
      "key": "cascader_{{test-id-nanoid}}",
      "label": "Cascader",
      "options": {
        "checkStrictly": false,
        "clearable": true,
        "data": [
          {
            "label": "Option1",
            "value": "{{test-id-nanoid}}",
          },
          {
            "label": "Option2",
            "value": "{{test-id-nanoid}}",
          },
          {
            "label": "Option3",
            "value": "{{test-id-nanoid}}",
          },
        ],
        "defaultValue": "",
        "disabled": false,
        "filterable": true,
        "isShowLabel": true,
        "labelWidth": 100,
        "multiple": false,
        "placeholder": "Please select",
        "required": false,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "cascader",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Cellphone 1`] = `
{
  "columns": [
    {
      "icon": "cellphone",
      "id": "{{test-id-nanoid}}",
      "key": "input_{{test-id-nanoid}}",
      "label": "Cellphone",
      "options": {
        "clearable": true,
        "defaultValue": "",
        "disabled": false,
        "isShowLabel": true,
        "labelWidth": 100,
        "placeholder": "Please enter",
        "renderType": 4,
        "required": false,
        "showPassword": false,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "input",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Checkbox 1`] = `
{
  "columns": [
    {
      "icon": "checkbox",
      "id": "{{test-id-nanoid}}",
      "key": "checkbox_{{test-id-nanoid}}",
      "label": "Checkbox",
      "options": {
        "data": [
          {
            "label": "Option1",
            "value": "{{test-id-nanoid}}",
          },
          {
            "label": "Option2",
            "value": "{{test-id-nanoid}}",
          },
          {
            "label": "Option3",
            "value": "{{test-id-nanoid}}",
          },
        ],
        "dataKey": "",
        "defaultValue": [],
        "disabled": false,
        "displayStyle": "block",
        "isShowLabel": true,
        "labelWidth": 100,
        "required": false,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "checkbox",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Date 1`] = `
{
  "columns": [
    {
      "icon": "calendar",
      "id": "{{test-id-nanoid}}",
      "key": "date_{{test-id-nanoid}}",
      "label": "Date",
      "options": {
        "clearable": true,
        "defaultValue": null,
        "disabled": false,
        "endTime": null,
        "format": "YYYY-MM-DD",
        "isShowLabel": true,
        "isShowWordLimit": false,
        "labelWidth": 100,
        "placeholder": "Please select",
        "required": false,
        "startTime": null,
        "type": "date",
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "date",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Email 1`] = `
{
  "columns": [
    {
      "icon": "email",
      "id": "{{test-id-nanoid}}",
      "key": "input_{{test-id-nanoid}}",
      "label": "Email",
      "options": {
        "clearable": true,
        "defaultValue": "",
        "disabled": false,
        "isShowLabel": true,
        "labelWidth": 100,
        "placeholder": "Please enter",
        "renderType": 2,
        "required": false,
        "showPassword": false,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "input",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Html 1`] = `
{
  "columns": [
    {
      "icon": "html",
      "id": "{{test-id-nanoid}}",
      "key": "html_{{test-id-nanoid}}",
      "label": "Html",
      "options": {
        "action": "",
        "defaultValue": "",
        "disabled": false,
        "isShowLabel": true,
        "labelWidth": 100,
        "placeholder": "Please enter",
        "required": false,
        "size": 1,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "html",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > ID number 1`] = `
{
  "columns": [
    {
      "icon": "ID",
      "id": "{{test-id-nanoid}}",
      "key": "input_{{test-id-nanoid}}",
      "label": "ID number",
      "options": {
        "clearable": true,
        "defaultValue": "",
        "disabled": false,
        "isShowLabel": true,
        "labelWidth": 100,
        "placeholder": "Please enter",
        "renderType": 3,
        "required": false,
        "showPassword": false,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "input",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Input 1`] = `
{
  "columns": [
    {
      "icon": "input",
      "id": "{{test-id-nanoid}}",
      "key": "input_{{test-id-nanoid}}",
      "label": "Input",
      "options": {
        "clearable": true,
        "defaultValue": "",
        "disabled": false,
        "isShowLabel": true,
        "isShowWordLimit": false,
        "labelWidth": 100,
        "max": null,
        "min": null,
        "placeholder": "Please enter",
        "renderType": 1,
        "required": false,
        "showPassword": false,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "input",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Number 1`] = `
{
  "columns": [
    {
      "icon": "numbers",
      "id": "{{test-id-nanoid}}",
      "key": "number_{{test-id-nanoid}}",
      "label": "Number",
      "options": {
        "controls": true,
        "controlsPosition": true,
        "defaultValue": null,
        "disabled": false,
        "isShowLabel": true,
        "labelWidth": 100,
        "max": null,
        "maxlength": null,
        "min": 0,
        "minlength": null,
        "precision": 0,
        "required": false,
        "step": 1,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "number",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Radio 1`] = `
{
  "columns": [
    {
      "icon": "radio",
      "id": "{{test-id-nanoid}}",
      "key": "radio_{{test-id-nanoid}}",
      "label": "Radio",
      "options": {
        "data": [
          {
            "label": "Option1",
            "value": "{{test-id-nanoid}}",
          },
          {
            "label": "Option2",
            "value": "{{test-id-nanoid}}",
          },
          {
            "label": "Option3",
            "value": "{{test-id-nanoid}}",
          },
        ],
        "dataKey": "",
        "defaultValue": "",
        "disabled": false,
        "displayStyle": "block",
        "isShowLabel": true,
        "labelWidth": 100,
        "required": false,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "radio",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Rate 1`] = `
{
  "columns": [
    {
      "icon": "rating",
      "id": "{{test-id-nanoid}}",
      "key": "rate_{{test-id-nanoid}}",
      "label": "Rate",
      "options": {
        "allowHalf": false,
        "defaultValue": 0,
        "disabled": false,
        "isShowLabel": true,
        "labelWidth": 100,
        "max": 5,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "rate",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Region 1`] = `
{
  "columns": [
    {
      "icon": "location",
      "id": "{{test-id-nanoid}}",
      "key": "region_{{test-id-nanoid}}",
      "label": "Region",
      "options": {
        "defaultValue": "",
        "filterable": true,
        "isShowLabel": true,
        "labelWidth": 100,
        "placeholder": "Please select",
        "required": false,
        "selectType": 3,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "region",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Select 1`] = `
{
  "columns": [
    {
      "icon": "dropdown0",
      "id": "{{test-id-nanoid}}",
      "key": "select_{{test-id-nanoid}}",
      "label": "Select",
      "options": {
        "clearable": true,
        "data": [
          {
            "label": "Option1",
            "value": "{{test-id-nanoid}}",
          },
          {
            "label": "Option2",
            "value": "{{test-id-nanoid}}",
          },
          {
            "label": "Option3",
            "value": "{{test-id-nanoid}}",
          },
        ],
        "dataKey": "",
        "defaultValue": "",
        "disabled": false,
        "filterable": true,
        "isShowLabel": true,
        "labelWidth": 100,
        "multiple": false,
        "placeholder": "Please select",
        "required": false,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "select",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Slider 1`] = `
{
  "columns": [
    {
      "icon": "slider",
      "id": "{{test-id-nanoid}}",
      "key": "slider_{{test-id-nanoid}}",
      "label": "Slider",
      "options": {
        "defaultValue": 0,
        "disabled": false,
        "isShowLabel": true,
        "labelWidth": 100,
        "max": 100,
        "min": 0,
        "placeholder": "",
        "required": false,
        "step": 1,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "slider",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Switch 1`] = `
{
  "columns": [
    {
      "icon": "switch",
      "id": "{{test-id-nanoid}}",
      "key": "switch_{{test-id-nanoid}}",
      "label": "Switch",
      "options": {
        "defaultValue": true,
        "disabled": false,
        "isShowLabel": true,
        "labelWidth": 100,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "switch",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Textarea 1`] = `
{
  "columns": [
    {
      "icon": "textarea",
      "id": "{{test-id-nanoid}}",
      "key": "textarea_{{test-id-nanoid}}",
      "label": "Textarea",
      "options": {
        "clearable": true,
        "defaultValue": "",
        "disabled": false,
        "isShowLabel": true,
        "isShowWordLimit": false,
        "labelWidth": 100,
        "max": null,
        "min": null,
        "placeholder": "Please enter",
        "required": false,
        "rows": 6,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "textarea",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > Time 1`] = `
{
  "columns": [
    {
      "icon": "time",
      "id": "{{test-id-nanoid}}",
      "key": "time_{{test-id-nanoid}}",
      "label": "Time",
      "options": {
        "clearable": true,
        "defaultValue": null,
        "disabled": false,
        "format": "HH:mm:ss",
        "isShowLabel": true,
        "labelWidth": 100,
        "placeholder": "Please select",
        "required": false,
        "valueFormat": "HH:mm:ss",
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "time",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;

exports[`Generate field data: > URL 1`] = `
{
  "columns": [
    {
      "icon": "url",
      "id": "{{test-id-nanoid}}",
      "key": "input_{{test-id-nanoid}}",
      "label": "URL",
      "options": {
        "clearable": true,
        "defaultValue": "",
        "disabled": false,
        "isShowLabel": true,
        "labelWidth": 100,
        "placeholder": "Please enter",
        "renderType": 5,
        "required": false,
        "showPassword": false,
      },
      "style": {
        "width": {
          "mobile": "100%",
          "pc": "100%",
        },
      },
      "type": "input",
    },
  ],
  "id": "{{test-id-nanoid}}",
  "key": "inline_{{test-id-nanoid}}",
  "style": {},
  "type": "inline",
}
`;
